stan/math/fwd/fun/multiply.hpp

```diff
@@ -10,120 +11,58 @@
 namespace stan {
 namespace math {
 
-template <typename T, int R1, int C1>
-inline Eigen::Matrix<fvar<T>, R1, C1> multiply(
-    const Eigen::Matrix<fvar<T>, R1, C1>& m, const fvar<T>& c) {
-  Eigen::Matrix<fvar<T>, R1, C1> res(m.rows(), m.cols());
-  for (int i = 0; i < m.size(); i++)
-    res(i) = c * m(i);
-  return res;
-}
-
-template <typename T, int R2, int C2>
-inline Eigen::Matrix<fvar<T>, R2, C2> multiply(
-    const Eigen::Matrix<fvar<T>, R2, C2>& m, double c) {
-  Eigen::Matrix<fvar<T>, R2, C2> res(m.rows(), m.cols());
-  for (int i = 0; i < m.size(); i++)
-    res(i) = c * m(i);
-  return res;
-}
-
-template <typename T, int R1, int C1>
-inline Eigen::Matrix<fvar<T>, R1, C1> multiply(
-    const Eigen::Matrix<double, R1, C1>& m, const fvar<T>& c) {
-  Eigen::Matrix<fvar<T>, R1, C1> res(m.rows(), m.cols());
-  for (int i = 0; i < m.size(); i++)
-    res(i) = c * m(i);
-  return res;
-}
-
-template <typename T, int R1, int C1>
-inline Eigen::Matrix<fvar<T>, R1, C1> multiply(
-    const fvar<T>& c, const Eigen::Matrix<fvar<T>, R1, C1>& m) {
-  return multiply(m, c);
-}
-
-template <typename T, int R1, int C1>
-inline Eigen::Matrix<fvar<T>, R1, C1> multiply(
-    double c, const Eigen::Matrix<fvar<T>, R1, C1>& m) {
-  return multiply(m, c);
-}
-
-template <typename T, int R1, int C1>
-inline Eigen::Matrix<fvar<T>, R1, C1> multiply(
-    const fvar<T>& c, const Eigen::Matrix<double, R1, C1>& m) {
-  return multiply(m, c);
-}
-
-template <typename T, int R1, int C1, int R2, int C2>
-inline Eigen::Matrix<fvar<T>, R1, C2> multiply(
-    const Eigen::Matrix<fvar<T>, R1, C1>& m1,
-    const Eigen::Matrix<fvar<T>, R2, C2>& m2) {
+template <typename T1, typename T2,
+          typename = require_all_eigen_vt<is_fvar, T1, T2>,
+          typename = require_same_t<typename value_type_t<T1>::Scalar,
+                                    typename value_type_t<T2>::Scalar>,
+          typename = require_not_t<
+              conjunction<is_eigen_row_vector<T1>, is_eigen_col_vector<T2>>>,
```

Comment:
[optional]
We could have something like `require_eigen_dot_product (I'm not sure if there is some other math word for general ops with a row and column vector.

Also I think this could use `require_any_not_t`

Reply:
Also names like RowVec and ColVec may be good here

Reply:
Here we have require_NOT, so these are not row and col vec.